apollo-federation/src/sources/connect/json_selection/location.rs

```diff
@@ -0,0 +1,372 @@
+use nom::bytes::complete::tag;
+use nom::combinator::map;
+use nom::IResult;
+use nom_locate::LocatedSpan;
+
+pub type Span<'a> = LocatedSpan<&'a str>;
+
+// Some parsed AST structures, like PathSelection and NamedSelection, can
+// produce a range directly from their children, so they do not need to be
+// wrapped as WithRange<PathSelection> or WithRange<NamedSelection>.
+// Additionally, AST nodes that are structs can store their own range as a
+// field, so they can implement Ranged<T> without the WithRange<T> wrapper.
+pub trait Ranged<T> {
+    fn node(&self) -> &T;
+    fn range(&self) -> OffsetRange;
+}
+
+// The ranges produced by the JSONSelection parser are pairs of character
+// offsets into the original string. The first element of the pair is the offset
+// of the first character, and the second element is the offset of the character
+// just past the end of the range. Offsets start at 0 for the first character in
+// the file, following nom_locate's span.location_offset() convention.
+pub type OffsetRange = Option<(usize, usize)>;
```

Comment:
Is there a reason this is encoded as a tuple versus the [built-in range type](https://doc.rust-lang.org/std/ops/struct.Range.html)?

Reply:
That's a great idea!

Reply:
https://github.com/apollographql/router/pull/5987/commits/774ae142dcb59a8699bf1c685a2d48c00bf56311